src/node_v8_platform.h

```diff
@@ -0,0 +1,184 @@
+#ifndef SRC_NODE_V8_PLATFORM_H_
+#define SRC_NODE_V8_PLATFORM_H_
+
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+
+#include <memory>
+
+#include "env-inl.h"
+#include "node.h"
+#include "node_metadata.h"
+#include "node_platform.h"
+#include "node_options.h"
+#include "tracing/node_trace_writer.h"
+#include "tracing/trace_event.h"
+#include "tracing/traced_value.h"
+#include "util.h"
+
+namespace node {
+
+// Ensures that __metadata trace events are only emitted
+// when tracing is enabled.
+class NodeTraceStateObserver
+    : public v8::TracingController::TraceStateObserver {
+ public:
+  void OnTraceEnabled() override {
+    std::string title = GetProcessTitle("");
+    if (!title.empty()) {
+      // Only emit the metadata event if the title can be retrieved
+      // successfully. Ignore it otherwise.
+      TRACE_EVENT_METADATA1(
+          "__metadata", "process_name", "name", TRACE_STR_COPY(title.c_str()));
+    }
+    TRACE_EVENT_METADATA1("__metadata",
+                          "version",
+                          "node",
+                          per_process::metadata.versions.node.c_str());
+    TRACE_EVENT_METADATA1(
+        "__metadata", "thread_name", "name", "JavaScriptMainThread");
+
+    auto trace_process = tracing::TracedValue::Create();
+    trace_process->BeginDictionary("versions");
+
+#define V(key)                                                                 \
+  trace_process->SetString(#key, per_process::metadata.versions.key.c_str());
+
+    NODE_VERSIONS_KEYS(V)
+#undef V
+
+    trace_process->EndDictionary();
+
+    trace_process->SetString("arch", per_process::metadata.arch.c_str());
+    trace_process->SetString("platform",
+                             per_process::metadata.platform.c_str());
+
+    trace_process->BeginDictionary("release");
+    trace_process->SetString("name",
+                             per_process::metadata.release.name.c_str());
+#if NODE_VERSION_IS_LTS
+    trace_process->SetString("lts", per_process::metadata.release.lts.c_str());
+#endif
+    trace_process->EndDictionary();
+    TRACE_EVENT_METADATA1(
+        "__metadata", "node", "process", std::move(trace_process));
+
+    // This only runs the first time tracing is enabled
+    controller_->RemoveTraceStateObserver(this);
+  }
+
+  void OnTraceDisabled() override {
+    // Do nothing here. This should never be called because the
+    // observer removes itself when OnTraceEnabled() is called.
+    UNREACHABLE();
+  }
+
+  explicit NodeTraceStateObserver(v8::TracingController* controller)
+      : controller_(controller) {}
+  ~NodeTraceStateObserver() override = default;
+
+ private:
+  v8::TracingController* controller_;
+};
+
+struct V8Platform {
+  bool initialized_ = false;
+
+#if NODE_USE_V8_PLATFORM
+  void Initialize(int thread_pool_size) {
```

Comment:
Method definitions like this - wherein a method is non-inline if `NODE_USE_V8_PLATFORM` but inline otherwise - have been intentionally kept in the `.h` file, as opposed to moving them to a `.cc` file.

The idea being - to prevent confusion and mistakes due to code duplication of the `NODE_USE_V8_PLATFORM` if-else construct, e.g.:
```cpp
// src/node_v8_platform.h

class NodeTraceStateObserver {

#if NODE_USE_V8_PLATFORM
  void Initialize(int thread_pool_size);
#else
  inline void Initialize(int thread_pool_size);
#endif // NODE_USE_V8_PLATFORM

}
```

```cpp
// src/node_v8_platform.cc

#if NODE_USE_V8_PLATFORM
void NodeTraceStateObserver::Initialize(int thread_pool_size) {
  // ...
}
#else
inline void NodeTraceStateObserver::Initialize(int thread_pool_size) {
  // ...
}
#endif // NODE_USE_V8_PLATFORM
```